docs/design/0012-generic-association.md

```diff
@@ -0,0 +1,275 @@
+
+In this option a new CRD is introduced. It allows an Elastic stack application owner to specify for some third party applications which association data is required.
+
+#### Association specification as a new CRD
+
+```yaml
+apiVersion: association.k8s.elastic.co/v1alpha1
+kind: Association
+metadata:
+  name: monitor
+  namespace: monitor-ns # the Elastic stack applicaitons must live in this namespace
+spec:
+
+  ## targetNamespaces are the namespaces where the association information are copied and reconciled
+  targetNamespaces:
+    - ns1
+
+  ## These are the Elastic applications.
+  ## The namespace is automatically inherited from this custom resource.
+  elasticsearchName: monitor
+  kibanaName: monitor
+  # apmServerName: monitor
+  
+  ## User field lets the user specify a user or a Secret which contains a user and a password.
+  ## 2 versions are maintained:
+  ## 1. One in the target namespaces, as a part of the association information, it contains the username and the clear-text password
+  ## 2. An other one in the Elasticsearch namespace, it contains the hashed password, ready to be used by Elasticsearch
+  ## If empty then no user is generated, only certificates are copied as part of this association
+  user:
+    ## Role must already exists.
+    ## See https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-users-and-roles.html#k8s_creating_custom_roles
+    role: superuser
+    ## Optional and proposed here as a later improvement: user may specify a spec of the password.
+    ## Some sane defaults would be: 64 chars, mixed cases, no special characters and no ttl.
+    # passwordSpec:
+    #   length: 64
+    #   mixedCases: true
+    #   symbols: false
+    #   ttl: "24h"
+    # Or just reference a password
+    # secretName: secret-with-user-and-password
+```
+
+#### Automatic injection of the association data
+
+The new CRD allows creating and maintaining a copy of the information needed to establish an association. But, it might be worth to propose a mechanism to inject the variables automatically. Optionally, it would be possible to add some annotations to automatically inject the required data into the `Pods` _(in the targeted namespaces)_ :
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: my-application
+  annotations:
+    association.k8s.elastic.co/name: "monitor-ns/monitor" # namespaced name of the association CR
+spec:
+  selector:
+    matchLabels:
+      app: test
+  template:
+    metadata:
+      labels:
+        app: test
+    spec:
+      containers:
+      - name: container1
+        image: ...
+      - name: container2
+        image: ...
+```
+
+## Making association variables available
+
+### The Association Secret
+
+The proposal here is to define the content of the association secret. This `Secret` is managed by a generic association controller. According to what the user specify in the generic association custom resource this `Secret` may be filled with the following items:
```

Comment:
This only works in situations where ECK is managing the target namespace.
vs. a situation where ECK manages resources in namespaceA but the end-user application runs in namespaceB.
In that case, a cluster-wide secret-sync operator, decoupled from ECK, could make more sense? (something like https://github.com/IBM-Cloud/kube-samples/tree/master/secret-sync-operator)

Reply:
True if we only want to share a Secret (and I agree it could be a first step). But if we want to update/mutate some resources we must manage the target namespace anyway.